src/handlers/psp.ts

```diff
@@ -289,30 +298,36 @@ const groupSubscriptionsBySponsorWallet = async (
       ]) as Promise<node.LogsData<(SponsorWalletTransactionCount | null) & { sponsor: string }>>
   );
   const sponsorWalletsAndTransactionCounts = await Promise.all(sponsorWalletAndTransactionCountPromises);
-  sponsorWalletsAndTransactionCounts.forEach(([logs, data]) => {
-    const sponsorLogOptions: node.LogOptions = {
-      ...providerLogOptions,
-      additional: {
-        ...providerLogOptions.additional,
-        sponsor: data.sponsor,
-      },
-    };
-    node.logger.logPending(logs, sponsorLogOptions);
+  const sponsorWalletsWithSubscriptions = sponsorWalletsAndTransactionCounts.reduce(
+    (acc: SponsorWalletWithSubscriptions[], [logs, data]) => {
+      const sponsorLogOptions: node.LogOptions = {
+        ...providerLogOptions,
+        additional: {
+          ...providerLogOptions.additional,
+          sponsor: data.sponsor,
+        },
+      };
+      node.logger.logPending(logs, sponsorLogOptions);
 
-    if (isNil(data.sponsorWallet) || isNil(data.transactionCount)) {
-      node.logger.warn('Failed to fetch sponsor wallet or transaction count', sponsorLogOptions);
-      return;
-    }
+      if (isNil(data.sponsorWallet) || isNil(data.transactionCount)) {
+        node.logger.warn('Failed to fetch sponsor wallet or transaction count', sponsorLogOptions);
+        return acc;
+      }
 
-    let nextNonce = data.transactionCount;
-    sponsorWalletsWithSubscriptions.push({
-      subscriptions: subscriptionsBySponsor[data.sponsor].map((subscription) => ({
-        ...subscription,
-        nonce: nextNonce++,
-      })),
-      sponsorWallet: data.sponsorWallet,
-    });
-  });
+      let nextNonce = data.transactionCount;
+      return [
+        ...acc,
+        {
+          subscriptions: subscriptionsBySponsor[data.sponsor].map((subscription) => ({
+            ...subscription,
+            nonce: nextNonce++,
```

Comment:
I don't know if this is an appropriate time to point this out or change this:
The `nextNonce` can reference the index of the `map`, like:
```
      return [
        ...acc,
        {
          subscriptions: subscriptionsBySponsor[data.sponsor].map((subscription, idx) => ({
            ...subscription,
            nonce: data.transactionCount + idx + 1,
          })),
```
this gets rid of the mutation.

Reply:
Seems reasonable :smile: ... fixed

Reply:
I'm not sure this would work tbh because first transaction for a wallet should have a nonce of 0 and here we would always start at 1 🤔

Reply:
@acenolaza Wow you're right, nice catch as always... ill fix

Reply:
Fixed here: https://github.com/api3dao/airkeeper/commit/0b4957d993b3c50f8624039c4bddab208e8b97d6

I ❤️ the usage of reduce btw